### PR TITLE
Fix for polymorphic fields in SqlMapping

### PR DESCRIPTION
**Pull request by milessabin**
+ The resolution of polymorphic field mappings now takes the runtime context into account properly.
+ All narrowing operations on cursors yield a value in `Result` allowing mappings to report internal errors for failed discriminators.
+ In SQL mappings, failed discriminators now fail the entire query with an internal error.
+ The `TypeCase` extractor has been removed because it can't take the context into account.

### CODE REVIEW

**Comment by tpolecat** [comment]
Can you summarize what this change does? It looks like it's causing all of our discriminator predicates to fail.

**Review by tpolecat — APPROVED** [review]
:+1: this is working fine for us. The problem was on our end and was related to one of the new static methods on `ResultT`, for which we had a local syntax method with a different type (everything continued to typecheck after the update but some database operations were no longer being executed)

**Comment by milessabin** [comment]
@tpolecat thanks for the update!

**Comment by milessabin** [comment]
Rebased.
